src/bootstrap/ci_profiler.rs

```diff
@@ -0,0 +1,216 @@
+}
+
+impl CiProfiler {
+    pub(crate) fn init() -> Self {
+        let (reset_send, reset_recv) = std::sync::mpsc::channel();
+
+        let state = Arc::new(Mutex::new(ProfilerState {
+            finished_steps: Vec::new(),
+            running_steps: Vec::new(),
+
+            cpu_single_core: Vec::new(),
+            cpu_all_cores: Vec::new(),
+            timer_start: None,
+
+            reset_cpu_check: reset_send,
+        }));
+
+        let state_cloned = state.clone();
+        std::thread::spawn(move || cpu_sampler_thread(state_cloned, reset_recv));
+
+        CiProfiler { state }
+    }
+
+    pub(crate) fn enter_step<S: Step>(&self, step: &S) {
+        let mut state = t!(self.state.lock());
+
+        // Consider all the stats gathered so far as the parent's.
+        if !state.running_steps.is_empty() {
+            self.collect_stats(&mut *state);
+        }
+
+        let _ = state.reset_cpu_check.send(());
+
+        state.cpu_single_core.clear();
+        state.cpu_all_cores.clear();
+        state.timer_start = Some(Instant::now());
+
+        state.running_steps.push(ProfiledStep {
+            type_: std::any::type_name::<S>().into(),
+            debug_repr: format!("{step:?}"),
+
+            duration: Duration::ZERO,
+            cpu_single_core: Vec::new(),
+            cpu_all_cores: Vec::new(),
+
+            childs: Vec::new(),
+        });
+    }
+
+    pub(crate) fn exit_step(&self) {
+        let mut state = t!(self.state.lock());
+
+        self.collect_stats(&mut *state);
+
+        let step = state.running_steps.pop().unwrap();
+        if state.running_steps.is_empty() {
+            state.finished_steps.push(step);
+            state.timer_start = None;
+        } else {
+            state.running_steps.last_mut().unwrap().childs.push(step);
+
+            // Resume the timer for the parent step.
+            state.timer_start = Some(Instant::now());
+        }
+    }
+
+    fn collect_stats(&self, state: &mut ProfilerState) {
+        let step = state.running_steps.last_mut().unwrap();
+
+        step.duration += state.timer_start.unwrap().elapsed();
+        step.cpu_single_core.append(&mut state.cpu_single_core);
+        step.cpu_all_cores.append(&mut state.cpu_all_cores);
+    }
+
+    pub(crate) fn persist(&self, build: &Build) {
+        let mut state = t!(self.state.lock());
+        assert!(state.running_steps.is_empty(), "steps are still executing");
+
+        let steps = std::mem::take(&mut state.finished_steps);
+        let json = vec![JsonNode::RustbuildInvocation {
+            childs: steps.into_iter().map(|step| self.prepare_json_step(step)).collect(),
+        }];
+
+        let dest = if let Some(dest) = std::env::var_os("RUSTBUILD_CI_PROFILES_DEST") {
+            dest.into()
+        } else {
+            build.out.join("ci-profiler-output.json")
+        };
+
+        t!(std::fs::create_dir_all(dest.parent().unwrap()));
+        let mut file = t!(File::create(&dest));
```

Comment:
Yeah. This was something I originally hoped to use the environment variable for, outputting the information into separate files, to avoid deserializing and serializing back everything. I think deserializing and merging is the best strategy here, I'll implement that.